go/6_auth-kubernetes.go

```diff
@@ -0,0 +1,92 @@
+package main
+
+import (
+	"fmt"
+	"os"
+
+	vault "github.com/hashicorp/vault/api"
+)
+
+// Fetches a key-value secret (kv-v2) after authenticating to Vault with a Kubernetes service account.
+//
+// As the client, all we need to do is pass along the JWT token representing our application's Kubernetes Service Account in our login request to Vault.
+// This token is automatically mounted to your application's container by Kubernetes. Read more at https://www.vaultproject.io/docs/auth/kubernetes
+//
+// SETUP NOTES: If an operator has not already set up Kubernetes auth in Vault for you, then you must also first configure the Vault server with its own Service Account token to be able to communicate with the Kubernetes API
+// so it can verify that the client's service-account token is valid. The service account that will be performing that verification needs the ClusterRole system:auth-delegator.
+//
+//    export TOKEN_REVIEW_JWT=$(kubectl get secret $TOKEN_REVIEWER_SECRET --output='go-template={{ .data.token }}' | base64 --decode)
+//    export KUBE_HOST=$(kubectl config view --raw --minify --flatten --output='jsonpath={.clusters[].cluster.server}')
+//    kubectl config view --raw --minify --flatten --output='jsonpath={.clusters[].cluster.certificate-authority-data}' | base64 --decode > path/to/kube_ca_cert
+//
+//    vault write auth/kubernetes/config \
+//  	token_reviewer_jwt=${TOKEN_REVIEW_JWT} \
+//      kubernetes_host=${KUBE_HOST} \
+//      kubernetes_ca_cert=@path/to/kube_ca_cert \
+//      issuer="kubernetes/serviceaccount"
+//
+// The "issuer" field only needs to be set when running Kubernetes version 1.21 or above:
+// https://www.vaultproject.io/docs/platform/k8s/csi#setting-issuer-for-kubernetes-authentication.
+//
+// Finally, make sure to create a role in Vault bound to your pod's service account:
+//
+// 	vault write auth/kubernetes/role/dev-role-k8s \
+//     	policies="dev-policy" \
+//     	bound_service_account_names="my-app" \
+//		bound_service_account_namespaces="default"
+func getSecretWithKubernetesAuth() (string, error) {
+	// If set, the VAULT_ADDR environment variable will be the address that your pod uses to communicate with Vault.
+	// For an application running in Kubernetes accessing a Vault server outside of Kubernetes, you will need to set VAULT_ADDR in your pod spec to
+	// point at your Kubernetes cluster's gateway address. See this tutorial for more information: https://learn.hashicorp.com/tutorials/vault/kubernetes-external-vault
```

Comment:
> For an application running in Kubernetes accessing a Vault server outside of Kubernetes, you will need to set VAULT_ADDR in your pod spec to point at your Kubernetes cluster's gateway address.

This is true for the learn guide setup, but in general VAULT_ADDR should just point to whatever address the external Vault is being served from. (I think the gateway part comes into play in the learn guide because of how Docker is configured, and is specific to how docker containers connect to services running outside of Docker on the host.)

Reply:
Ohhhh okay, gotcha. I'll remove that part of the comment then, thank you.